### PR TITLE
Fix NPEs in Distribute

### PR DESCRIPTION
**Pull request by Jamminroot**
Please have a look at our [guidelines for contributions](https://github.com/microsoft/appcenter-sdk-android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

This PR fixes NPE in edge case in Distribute module

## Related PRs or issues

[AB#85361](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/85361)
#1477

### CODE REVIEW

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 1]
```diff
@@ -1621,6 +1627,12 @@ public void onClick(DialogInterface dialog, int which) {
     @UiThread
     private synchronized void showUpdateSetupFailedDialog() {
 
+        /* Check if we are in foreground. */
```
Comment (DmitriyKirakosyan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1630]:
```suggestion
        /* Check if we are in background. */
```

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 2]
```diff
@@ -1859,22 +1871,30 @@ private synchronized void showDownloadProgress() {
...
-            dialogBuilder.setPositiveButton(R.string.appcenter_distribute_install, new DialogInterface.OnClickListener() {
 
-                @Override
-                public void onClick(DialogInterface dialog, int which) {
-                    installMandatoryUpdate(releaseDetails);
-                }
-            });
-            mCompletedDownloadDialog = dialogBuilder.create();
-            showAndRememberDialogActivity(mCompletedDownloadDialog);
+        /* Check if we are in foreground. */
```
Comment (DmitriyKirakosyan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]:
```suggestion
        /* Check if we are in background. */
```

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 3]
```diff
@@ -1859,22 +1871,30 @@ private synchronized void showDownloadProgress() {
...
+        if (!shouldRefreshDialog(mCompletedDownloadDialog)) {
+            return;
+        }
+        final ReleaseDetails releaseDetails = mReleaseDetails;
+        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(mForegroundActivity);
+        dialogBuilder.setCancelable(false);
+        dialogBuilder.setTitle(R.string.appcenter_distribute_install_ready_title);
+        dialogBuilder.setMessage(getInstallReadyMessage());
+        dialogBuilder.setPositiveButton(R.string.appcenter_distribute_install, new DialogInterface.OnClickListener() {
+
```
Comment (DmitriyKirakosyan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1890]: Extra empty line

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 2 above)
Comment (Jamminroot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: Nope. I'd rather change it to `Verify we are in foreground` (both comments)

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 3 above)
Comment (DmitriyKirakosyan) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1890]: Or is it expected?

Comment (Jamminroot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1890]:
```suggestion
```

**Review by DmitriyKirakosyan — DISMISSED** [review]

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 4]
```diff
@@ -1859,22 +1871,29 @@ private synchronized void showDownloadProgress() {
...
-            dialogBuilder.setPositiveButton(R.string.appcenter_distribute_install, new DialogInterface.OnClickListener() {
 
-                @Override
-                public void onClick(DialogInterface dialog, int which) {
-                    installMandatoryUpdate(releaseDetails);
-                }
-            });
-            mCompletedDownloadDialog = dialogBuilder.create();
-            showAndRememberDialogActivity(mCompletedDownloadDialog);
+        /* Check if we are in foreground. */
```
Comment (AnastasiaKubova) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: maybe background?

Comment (AnastasiaKubova) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: background?

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 ### App Center Distribute
 
+* **[Fix]** Fix NullPointerExceptions occurring in Distribute module.
```
Comment (AnastasiaKubova) [inline CHANGELOG.md:7]: Maybe make sense to clarify that this happens only when Dialog should be shown?

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 2 above)
Comment (AnastasiaKubova) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: Agree with Dmitriy, I think that `background` will be more correct.

Comment (Jamminroot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: Nope.
It's either:
Confirm\check\verify we are **not** in the background
Or
Confirm\check\verify we are in the foreground

Comment (Jamminroot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: Or at least, something like "Return if we are in the background\not in the foreground" at least, as a compromise :D

Comment (Jamminroot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1875]: I'll adjust message then

File: CHANGELOG.md (hunk 5 above)
Comment (Jamminroot) [reply CHANGELOG.md:7]: If only "dialogs should be shown", as specifying exact cases would give too much unnecessary details. Thanks

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 6]
```diff
@@ -1621,6 +1627,12 @@ public void onClick(DialogInterface dialog, int which) {
     @UiThread
     private synchronized void showUpdateSetupFailedDialog() {
 
+        /* Do not attempt to show dialog if application is in the background. */
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1630]: Did we test that resuming app restore the flow at the intended step?

Comment (Jamminroot) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java:1630]: I see no reasons why it should not work. As a scope of the WI I've worked on NPEs on the first place, and If the behaviour is flawed (as in: not resumed), I'd better create a separate WI to address that.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeCustomizationTest.java [hunk 7]
```diff
@@ -593,6 +593,66 @@ public void handleUserUpdateActionDownloadForOptionalUpdate() throws Exception {
...
+        /* Start Distribute service. */
+        start(distribute);
+        distribute.onActivityResumed(mActivity);
+        distribute.onActivityPaused(mActivity);
+
+        /* Call handleUpdateAction. */
+        when(InstallerUtils.isUnknownSourcesEnabled(eq(mContext))).thenReturn(false);
+        distribute.handleUpdateAction(UpdateAction.UPDATE);
+
+        /* Verify UPDATE has been processed. */
```
Comment (Jamminroot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeCustomizationTest.java:621]:
```suggestion
        /* Verify UPDATE has not been processed. */
```

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeCustomizationTest.java [hunk 8]
```diff
@@ -593,6 +593,66 @@ public void handleUserUpdateActionDownloadForOptionalUpdate() throws Exception {
...
+        distribute.handleUpdateAction(UpdateAction.UPDATE);
+
+        /* Verify UPDATE has not been processed. */
+        verify(distribute).enqueueDownloadOrShowUnknownSourcesDialog(any(ReleaseDetails.class));
+        verify(mDialogBuilder, never()).create();
+    }
+
+    @Test
+    public void doShowUnknownSourcesDialogWhileInForeground() throws Exception {
+        /* Mock. */
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeCustomizationTest.java:628]:
```suggestion

        /* Mock. */
```

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeCustomizationTest.java [hunk 9]
```diff
@@ -593,6 +593,66 @@ public void handleUserUpdateActionDownloadForOptionalUpdate() throws Exception {
         verify(distribute).enqueueDownloadOrShowUnknownSourcesDialog(any(ReleaseDetails.class));
     }
 
+    @Test
+    public void doNotShowUnknownSourcesDialogWhileInBackground() throws Exception {
+        /* Mock. */
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeCustomizationTest.java:598]:
```suggestion

        /* Mock. */
```

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java [hunk 10]
```diff
@@ -471,6 +472,65 @@ public void showDownloadProgressTest() {
...
+    @Test
+    public void doNotShowUpdateSetupFailedDialogInBackground() {
+        when(SharedPreferencesManager.getString(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY)).thenReturn("failed_message");
+
+        /* Trigger call. */
+        start();
+
+        /* Verify dialog. */
+        verify(mDialogBuilder, never()).create();
+        SharedPreferencesManager.remove(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java:516]: Shouldn't be the dialog shown the next time we resume the app? If the state is deleted then the flow would just reset? This should be tested to be sure what is the current behavior if and if that behavior is ok (can force a npe in the code just to test).

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java [hunk 11]
```diff
@@ -471,6 +472,65 @@ public void showDownloadProgressTest() {
...
+        Distribute.getInstance().onActivityResumed(mActivity);
+        start();
+
+
+        /* Verify dialog. */
+        verify(mDialogBuilder).create();
+        SharedPreferencesManager.remove(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY);
+    }
+
+
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java:533]:
```suggestion
```

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java (hunk 10 above)
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java:516]: I think it's missing verifyStatic(never()) before checking remove

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java [hunk 12]
```diff
@@ -471,6 +472,65 @@ public void showDownloadProgressTest() {
...
+        when(SharedPreferencesManager.getString(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY)).thenReturn("failed_message");
+
+        /* Trigger call. */
+        Distribute.getInstance().onActivityResumed(mActivity);
+        start();
+
+
+        /* Verify dialog. */
+        verify(mDialogBuilder).create();
+        SharedPreferencesManager.remove(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY);
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/DistributeTest.java:530]: verifyStatic(never()) ?

File: CHANGELOG.md [hunk 13]
```diff
@@ -8,6 +8,7 @@
 
 ### App Center Distribute
 
+* **[Fix]** Fix NullPointerExceptions occurring in Distribute module when dialogs were intended to be shown.
```
Comment (MatkovIvan) [inline CHANGELOG.md:11]:
```suggestion
* **[Fix]** Fix `NullPointerException`s occurring when dialogs were intended to be shown, but there is no foreground activity at that moment.
```

**Review by guperrot — DISMISSED** [review]

**Review by MatkovIvan — APPROVED** [review]
